sftp/sftp_test.go

```diff
@@ -0,0 +1,274 @@
+package sftp
+
+import (
+	"fmt"
+	"net"
+	"os"
+	"path/filepath"
+	"strconv"
+	"testing"
+	"time"
+
+	"github.com/golang/mock/gomock"
+	"github.com/ory/dockertest/v3"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/rudderlabs/rudder-go-kit/sftp/mock_sftp"
+	"github.com/rudderlabs/rudder-go-kit/testhelper/docker/resource/sshserver"
+)
+
+func TestSSHClientConfig(t *testing.T) {
+	// Read private key
+	privateKey, err := os.ReadFile("testdata/ssh/test_key")
+	require.NoError(t, err)
+
+	type testCase struct {
+		description   string
+		config        *SSHConfig
+		expectedError error
+	}
+
+	testCases := []testCase{
+		{
+			description:   "WithNilConfig",
+			config:        nil,
+			expectedError: fmt.Errorf("config should not be nil"),
+		},
+		{
+			description: "WithEmptyHostName",
+			config: &SSHConfig{
+				HostName:   "",
+				Port:       22,
+				User:       "someUser",
+				AuthMethod: "passwordAuth",
+				Password:   "somePassword",
+			},
+			expectedError: fmt.Errorf("hostname should not be empty"),
+		},
+		{
+			description: "WithEmptyPort",
+			config: &SSHConfig{
+				HostName:   "someHostName",
+				User:       "someUser",
+				AuthMethod: "passwordAuth",
+				Password:   "somePassword",
+			},
+			expectedError: fmt.Errorf("port should not be empty"),
+		},
+		{
+			description: "WithPassword",
+			config: &SSHConfig{
+				HostName:   "someHostName",
+				Port:       22,
+				User:       "someUser",
+				AuthMethod: "passwordAuth",
+				Password:   "somePassword",
+			},
+			expectedError: nil,
+		},
+		{
+			description: "WithPrivateKey",
+			config: &SSHConfig{
+				HostName:   "someHostName",
+				Port:       22,
+				User:       "someUser",
+				AuthMethod: "keyAuth",
+				PrivateKey: string(privateKey),
+			},
+			expectedError: nil,
+		},
+		{
+			description: "WithUnsupportedAuthMethod",
+			config: &SSHConfig{
+				HostName:   "HostName",
+				Port:       22,
+				User:       "someUser",
+				AuthMethod: "invalidAuth",
+				PrivateKey: "somePrivateKey",
+			},
+			expectedError: fmt.Errorf("unsupported authentication method"),
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.description, func(t *testing.T) {
+			sshConfig, err := sshClientConfig(tc.config)
+			if tc.expectedError != nil {
+				assert.EqualError(t, tc.expectedError, err.Error())
+				assert.Nil(t, sshConfig)
+			} else {
+				assert.Nil(t, err)
+				assert.NotNil(t, sshConfig)
+			}
+		})
+	}
+}
+
+func TestMain(t *testing.T) {
```

Comment:
`TestMain` can only accept `*testing.M` and it is executed for using a custom setup/teardown common for all other tests so this is going to be executed before everything else to prepare for the other tests. I think this isn't what you're trying to do here so I suggest renaming the function.